Docs: update the TracingPolicy reference for arm64 (#884)
cilium/tetragon: docs/content/en/docs/reference/tracing-policy.md

depend on your build.

Conveniently, you can list all kernel symbols reading the `/proc/kallsyms`
file. For example to search for the `write` syscall kernel function, you can
execute `sudo grep sys_write /proc/kallsyms`, the output should be similar to
this, minus the architecture specific prefixes.

```
ffffdeb14ea712e0 T __arm64_sys_writev
ffffdeb14ea73010 T ksys_write
ffffdeb14ea73140 T __arm64_sys_write
ffffdeb14eb5a460 t proc_sys_write
ffffdeb15092a700 d _eil_addr___arm64_sys_writev
ffffdeb15092a740 d _eil_addr___arm64_sys_write
```

You can see that the exact name of the symbol for the write syscall on our
kernel version is `__arm64_sys_write`. Note that on `x86_64`, the prefix should
be `__x64_` instead of `__arm64_`.

{{< caution >}}
Kernel symbols contain an architecture specific prefix when they refer to
syscall symbols. To write portable tracing policies, i.e. policies that can run
both on `x86_64` and `arm64`, just use the symbol name without the prefix.

For example, instead of writing `call: "__arm64_sys_write"` or `call:
"__x64_sys_write"`, just write `call: "sys_write"`, Tetragon will adapt and add

> the correct prefix based on the architecture of the underlying machine.

Reviewer: ```suggestion
the correct prefix based on the architecture of the underlying machine.
Note that the event generated as output currently includes the prefix.
```

{{< /caution >}}

In our example, we will explore a `kprobe` hooking into the
[`fd_install`](https://elixir.bootlin.com/linux/v6.1.9/source/fs/file.c#L602)
kernel function. The `fd_install` kernel function is called each time a file
descriptor is installed into the file descriptor table of a process, typically
referenced within system calls like `open` or `openat`. Hooking `fd_install`

> has its benefits and limitation, which are out of the scope of this document.

Reviewer: ```suggestion
has its benefits and limitations, which are out of the scope of this document.
```

the moment. It describes the arbitrary event in the kernel that will be traced.

##### Kprobes

Kprobes enables you to dynamically break into any kernel routine and collect
debugging and performance information non-disruptively. kprobes are highly
tight to your kernel version and might not be portable since the kernel symbols
depend on your build.

Conveniently, you can list all kernel symbols reading the `/proc/kallsyms`
file. For example to search for the `write` syscall kernel function, you can
execute `sudo grep sys_write /proc/kallsyms`, the output should be similar to
this, minus the architecture specific prefixes.

```
ffffdeb14ea712e0 T __arm64_sys_writev
ffffdeb14ea73010 T ksys_write
ffffdeb14ea73140 T __arm64_sys_write
ffffdeb14eb5a460 t proc_sys_write
ffffdeb15092a700 d _eil_addr___arm64_sys_writev
ffffdeb15092a740 d _eil_addr___arm64_sys_write
```

You can see that the exact name of the symbol for the write syscall on our
kernel version is `__arm64_sys_write`. Note that on `x86_64`, the prefix should
be `__x64_` instead of `__arm64_`.

{{< caution >}}
Kernel symbols contain an architecture specific prefix when they refer to

> syscall symbols. To write portable tracing policies, i.e. policies that can run

Reviewer 2: ```suggestion
syscall symbols. It's probably a good idea to write portable tracing policies that can run
```

Reviewer 3: Should this be resolved @mtardy ?

Author: I don't agree with the suggestion personally but we can debate

Reviewer 3: both are fine, so let's just merge as it is, and we can improve anyway later! much appreciated @errordeveloper for the feedback

Marked resolved

> both on `x86_64` and `arm64`, just use the symbol name without the prefix.

Reviewer 2: Won't the same apply to other architectures?

Author: Theoretically yes but Tetragon only supports amd64 and arm64 at the moment. It will need some manual changes to support a new arch. Do you have a specific arch in mind?

Reviewer 2: No, I'm just thinking that if sentence doesn't need to be so specific, e.g.

```suggestion
on any architecture, just use the symbol name without the architecture prefix.
```

Author: Yes but the reality is that we support only amd64 and arm64 so I'm not sure being implicit about supporting everything would help?

Reviewer 2: I guess it's just a bit more pragmatic, so you don't have to worry about this nuanced documentation goes out of date. But also, as user I wonder if other architectures maybe supported in the future or this policies will only ever work for the two architectures, which is a side effect of  such explicit docs.

Author: Ok I changed that to more generic `to run on multiple architectures`

Author: Ok I changed that to more generic `to run on multiple architectures`

    syscall: false
```

{{< note >}}
Notice the `syscall` field, specific to a `kprobe` spec, with default value
`false`, that indicates whether Tetragon will hook a syscall or just a regular
kernel function. Tetragon needs this information because syscall and kernel
function use a [different ABI](https://gitlab.com/x86-psABIs/x86-64-ABI).
{{< /note >}}

> Like usually, kprobes calls can be defined independently in different policies,

Reviewer: ```suggestion
As usual, kprobes calls can be defined independently in different policies,
```

enable        iocost          oom             skb           xdp
error_report  iomap           page_isolation  smbus         xhci-hcd
```

You can then choose the subsystem that you want to trace, and look the
tracepoint you want to use and its format. For example, if we choose the
`netif_receive_skb` tracepoints from the `net` subsystem, we can read its
format with `sudo cat /sys/kernel/debug/tracing/events/net/netif_receive_skb/format`,
the output should be similar to the following.

```
name: netif_receive_skb
ID: 1398
format:
        field:unsigned short common_type;       offset:0;       size:2; signed:0;
        field:unsigned char common_flags;       offset:2;       size:1; signed:0;
        field:unsigned char common_preempt_count;       offset:3;       size:1; signed:0;
        field:int common_pid;   offset:4;       size:4; signed:1;

        field:void * skbaddr;   offset:8;       size:8; signed:0;
        field:unsigned int len; offset:16;      size:4; signed:0;
        field:__data_loc char[] name;   offset:20;      size:4; signed:0;

print fmt: "dev=%s skbaddr=%px len=%u", __get_str(name), REC->skbaddr, REC->len
```

> 

Reviewer: ```suggestion
Similarly to kprobes, tracepoints can also hook into system calls. For more details, see the `raw_syscalls` and `syscalls` subysystems.
```
